app/src/main/java/edu/mills/heartoftaiwanese/storage/TranslationDatabaseHelper.kt

```diff
@@ -0,0 +1,253 @@
+        val translationValues = contentValuesOf(
+            tKeyChinese to word.chinese,
+            tKeyEnglish to word.english,
+            tKeyTaiwanese to word.taiwanese,
+            tKeyFavorite to 0,
+            tKeyAccessTime to currentTime
+        )
+        return writableDatabase.use {
+            it.insert(tTranslations, null, translationValues) != -1L
+        }
+    }
+
+    /**
+     * Gets the 5 most recently added translations, or all if there are 5 or less.
+     * @return a list of the newest (most recently added) words
+     */
+    fun getNewest(): List<DatabaseWord> {
+        return readableDatabase.use { db ->
+            db.query(
+                tTranslations, allColumns, null, null, null, null, "$tId DESC", 5.toString()
+            ).use { convertToDatabaseWord(it) }
+        }
+    }
+
+    // Gets the 5 most recently viewed translations, or all if there are 5 or less.
+    fun getRecent(): List<DatabaseWord> {
+        return readableDatabase.use { db ->
+            db.query(
+                tTranslations, allColumns, null, null,
+                null, null, "$tKeyAccessTime DESC", 5.toString()
+            ).use {
+                val wordList = convertToDatabaseWord(it)
+                readableDatabase.close()
+                wordList
+            }
+        }
+    }
+
+    /**
+     * @return all of the entries in which "favorites" is true.
+     */
+    fun getAllFavorites(): List<DatabaseWord> {
+        return readableDatabase.use { db ->
+            db.query(
+                tTranslations, allColumns, "$tKeyFavorite = ?", arrayOf("1"),
+                null, null, "$tKeyAccessTime DESC"
+            ).use {
+                val wordList = convertToDatabaseWord(it)
+                readableDatabase.close()
+                wordList
+            }
+        }
+    }
+
+    fun getWordByTaiwanese(taiwanese: String): DatabaseWord? {
+        return readableDatabase.use { db ->
+            db.query(
+                tTranslations, allColumns, "$tKeyTaiwanese = ?",
+                arrayOf(taiwanese), null, null, null, 1.toString()
+            ).use { convertToDatabaseWord(it).firstOrNull() }
+        }
+    }
+
+    fun getWordByEnglish(english: String): DatabaseWord? {
+        return readableDatabase.use { db ->
+            db.query(
+                tTranslations,
+                allColumns,
+                "$tKeyEnglish = ?",
+                arrayOf(english),
+                null,
+                null,
+                null,
+                1.toString()
+            ).use { convertToDatabaseWord(it).firstOrNull() }
+        }
+    }
+
+    fun getWordByChinese(chinese: String): DatabaseWord? {
+        return readableDatabase.query(
+            tTranslations, allColumns, "$tKeyChinese = ?",
+            arrayOf(chinese), null, null, null, 1.toString()
+        ).use {
+            val wordList = convertToDatabaseWord(it).firstOrNull()
+            readableDatabase.close()
```

Comment:
Should you even be closing the database? https://stackoverflow.com/questions/6608498/best-place-to-close-database-connection/7739454#7739454

It looks like you are inconsistent in doing so.

Reply:
you're right, I removed all of the `close()`